tools/mason/MasonPublish.chpl

```diff
@@ -0,0 +1,190 @@
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+use MasonUtils;
+use Spawn;
+use FileSystem;
+use TOML;
+use MasonEnv;
+use MasonNew;
+use MasonModify;
+
+/* Top Level procudeure that gets called from mason.chpl that takes in arguments from command line
+Retruns the help output in '-h' or '--help' exits in the arguements.
+If --dry-run is passed then it checks to see if the package is able to be published.
+Takes in the username of the package owner as an arguemnt*/
+proc masonPublish(args: [] string) throws {
+  try! {
+    if args.size == 2 {
+      masonPublishHelp();
+      exit(0);
+    }
+    var dry = false;
+    var username = "";
+    if args.size > 2 {
+      for arg in args[2..] {
+        if arg == '-h' || arg == '--help'{
+          masonPublishHelp();
+          exit(0);
+        }
+        else if arg == "--dry-run" {
+          dry = true;
+        }
+        else {
+          username = arg;
+        }
+      }
+      if doesGitOriginExist() {
+        if dry {
+          dryRun(username);
+        }
+        else {
+          publishPackage(username);
+        }
+      }
+      else {
+        throw new owned MasonError('Must have package set up as git repo to publish');
+      }
+    }
+  }
+  catch e: MasonError {
+    writeln(e.message());
+    exit(1);
+  }
+}
+
+/* Main Script that goes through the act of publishing the package to the mason registry.
+ Takes the package owners GitHub usrename as input will throw errors through command
+line git commands if any of the git calls fails.*/
+proc publishPackage(username: string) throws {
+  cloneMasonReg(username);
+  branchMasonReg(username);
+  const cwd = getEnv("PWD");
+  const package=  addPackageToBricks();
+  here.chdir(cwd + "/mason-registry");
+  const name = getName();
+  const url = gitUrl();
+  const projectHome = getProjectHome(cwd);
+  here.chdir(cwd +"/mason-registry/");
+  runCommand("git add .");
+  runCommand("git commit -m '" + package +"'");
+  runCommand('git push --set-upstream origin ' + package, true);
+  runCommand('git remote add upstream https://github.com/chapel-lang/mason-registry');
+  runCommand('git request-pull upstream/master https://github.com/chapel-lang/mason-registry ', true);
+  here.chdir(cwd);
+  rmTree('mason-registry');
+}
+
+/* If --dry-run is passed then it takes the usernane and checks to see if the mason-registry is forked
+ and the package has a git remote origin. If both exist then the package can be published. */
+proc dryRun(username: string) throws {
```

Comment:
Typically a `--dry-run` flag will also emit the commands it is going to run for the real run under the hood as well, such as the git commands.